src/orange/OrangeParams.cc

```diff
@@ -95,6 +95,65 @@ OrangeInput input_from_json(std::string filename)
     return result;
 }
 
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate the maximum num of universe levels within the geometry
+ *
+ * A single-universe geometry will have a max_num_levels of 1.
+ * This is a recursive function, so for external callers, uid should be
+ * the root universe id.
+ */
+size_type calc_max_num_levels(HostVal<OrangeParamsData> const& data,
+                              UniverseId const& uid)
+{
+    CELER_EXPECT(uid);
+
+    // Depth of the deepest daughter of uid
+    size_type max_sub_depth = 0;
+
+    if (data.universe_types[uid] == UniverseType::simple)
+    {
+        auto const& simple_unit_record
+            = data.simple_units[SimpleUnitId{data.universe_indices[uid]}];
+        for (auto vol_id : range(simple_unit_record.volumes.size()))
+        {
+            auto vol_record
+                = data.volume_records[simple_unit_record
+                                          .volumes[LocalVolumeId{vol_id}]];
+            if (vol_record.daughter_id)
+            {
+                max_sub_depth = std::max(
+                    max_sub_depth,
+                    calc_max_num_levels(
```

Comment:
This is probably fine for now, but in the future we'll want to keep a `map<UniverseId, size_type>` of the visited daughters so that we don't have to visit 1M pin cells: this would be refactored as
```
struct MaxNumLevelCalculator
{
    HostVal<OrangeParamsData> const& data;
    std::map<UniverseId, int> num_levels;
    int operator()(UniverseId);
};

Reply:
Ah that's a good point --- easy refactor